### ISSUE TITLE
Workflows and Activities should be able to access claims/permissions and other fields from validated JWT tokens

### ISSUE THREAD

**Issue by shaunco** [temporalio/temporal#1135]
**Is your feature request related to a problem? Please describe.**
Workflows and Activities need to be able to read claims/permissions and other fields from validated JWT tokens

**Describe the solution you'd like**
When using the authorizer and claimMappers, workflows and activities should be able to obtain the parsed and validated JWT, claims/permissions contained in the JWT, and the original bearer token - perhaps through `workflow.Context` in workflows and `context.Context` in activities

**Describe alternatives you've considered**
Our current implementation uses a propagator to pass these fields along from client to workflow to activity, but it is clunky and re-validating the token in the workflow is less than ideal (needing to fetch JWKS, side effects, and all). The new `authorizer` and `claimMappers` look great, but each workflow/activity should be able to reuse the provided token for calling other services, executing child workflows, and executing activities and should also be able to obtain/check/use claims and other token fields (like `sub` for Auth0 tokens).

**Comment by sergeybykov** [comment]
The authorization interceptor already [adds mapped claims to the context](https://github.com/temporalio/temporal/blob/77acf2c4699e33e51ebf02035820f9284e767024/common/authorization/interceptor.go#L105) so that downstream components can leverage information about Temporal claims associated with the caller.

Also, `Claims` struct includes a `Subject` field that allows claim mapper to put an arbitrary string there. So, you could put a Bearer token and any other information into with a custom claim mapper. Does that allow you to do what you need, even if indirectly?

We could consider adding raw Bearer token explicitly into context. My concern about doing that is that it would impose a performance tax on everybody, even customers that don't need it, or would require yet another config switch to turn the behavior on or off.

In the current implementation, claim mapper only cares about permission claims in the JWT token and ignores everything else. One potential option I see here it to allow for claim mappers to add raw claims in addition to handling permission claims. Would that solve your problem and allow to keep validation and parsing of token in one place - in the claim mapper? I also see a counter-argument that this sounds like a feature creep with claim mapper functionality growing beyond what was originally intended - to enable authorization.

**Comment by shaunco** [comment]
> The authorization interceptor already [adds mapped claims to the context](https://github.com/temporalio/temporal/blob/77acf2c4699e33e51ebf02035820f9284e767024/common/authorization/interceptor.go#L105) so that downstream components can leverage information about Temporal claims associated with the caller.

Thanks, I had totally missed this!

> Also, `Claims` struct includes a `Subject` field that allows claim mapper to put an arbitrary string there. So, you could put a Bearer token and any other information into with a custom claim mapper. Does that allow you to do what you need, even if indirectly?

Depends on the disposition of #1134. If we end up building our own server and containers, this will work. 

> We could consider adding raw Bearer token explicitly into context. My concern about doing that is that it would impose a performance tax on everybody, even customers that don't need it, or would require yet another config switch to turn the behavior on or off.

I'm not sure what Temporal's plans for the official docker images are, but if that is a route that is to continue forward, then having config to access built-in functionality seems necessary. I don't think anyone has ever complained that a server is too configurable...

**Comment by sergeybykov** [comment]
I commented in #1134 that we do want to support official docker images as much as possible. With that in mind, there are still multiple options here:
* Add Bearer token to context as is
* Have JWT claim mapper extract all claims and add non-permission ones to context
* Do both things

What would be your preference here?

**Comment by shaunco** [comment]
My vote would be for all aspects of the parsed and validated token to be available to workflows and activities. It seems like any omission of [public claims](https://www.iana.org/assignments/jwt/jwt.xhtml) or private claims (including `permissions`) is going to end up with someone asking for them. That is, I believe, your `both` option.

From a security and performance perspective, I'd rather see _only_ the original auth passed in the server headers to the client, and the client re-parse and re-validate it before activating the workflow or activity. With much of JWT moving to asymmetric (RSA or ECDSA based) tokens, this provides the best security posture, as the token can't be tampered with in Temporal's backing data store, and each workflow/activity knows that the token has been locally validated before execution rather than believing already parsed claims/permissions that arrived in headers. It also means that long lived workflows can fail if the original caller's authentication/authorization is revoked or expired. ... the catch to this is that it means authorizer and claimMappers need to be on both the server side and on the worker side to prevent trust from becoming transitive.

**Comment by sergeybykov** [comment]
> From a security and performance perspective, I'd rather see only the original auth passed in the server headers to the client, and the client re-parse and re-validate it before activating the workflow or activity. 

I was thinking along similar lines that passing just the original token might be cleaner and cheaper performance wise. My concern is that even if we were to pass all claims, somebody might ask for signature or something. That leads me to option 1, which would allow client to implement any scenario it needs while leaving the claim mapper / authorizer responsible only for authorizing calls to Temporal frontends.

**Comment by shaunco** [comment]
Agreed on all points.

> while leaving the claim mapper / authorizer responsible only for authorizing calls to Temporal frontends

It would be really nice if the client could pass the same authorizers and claim mappers to `client.NewClient` via `client.ConnectionOptions` in the sdk-go, as the structure and JWT code you put together is pretty nice, and having all JWT clients roll their own would be a shame... but any extensible mechanism that can be passed to `NewClient()`, `RegisterWorkflow()`, or `RegisterActivity()` would be great if the raw token arrives in a header.

That said, and a total tangent from above, we recently switched from jwt-go to https://github.com/lestrrat-go/jwx while implementing full JWKS support, as it is a *complete* JWT/JWK/JWA/JWS/JWE implementation, and is fantastically maintained and supported. JWKS support is as simple as:
```go
set, err := jwk.FetchHTTP(uri)
token, err := jwt.ParseString(
	tokenStr,
	jwt.WithKeySet(set),
)
jwt.Validate(token, opts...)
```

Claim/audience/clock/skew/issuer/subject/jwtID/ validation is all built-in via options.

**Comment by sergeybykov** [comment]
Interesting. We haven't considered reusing the authorization components on the client side. Let us take a look into it. But I think it deserves to be tracked by a separate issue. We discussed creating a new repo for code shared between server and client, but haven't started that yet.

Thanks for the pointer to https://github.com/lestrrat-go/jwx. I'll check it out.

**Comment by sergeybykov** [comment]
I submitted #1144. It passes the authorization header rather than just the token, for symmetry with what claim mapper expect.

**Comment by shaunco** [comment]
Fantastic, thank you!

**Comment by sergeybykov** [comment]
Resolved via #1144.
